apiserver/upgradeservice/upgradeimpl.go

```diff
@@ -114,31 +151,152 @@ func CreateUpgrade(request *msgs.CreateUpgradeRequest, ns string) msgs.CreateUpg
+			return response
 		}
-		if imageToUpgradeTo == cl.Spec.CCPImageTag {
+
+		// for the upgrade procedure, we only upgrade to the current image used by the
+		// Postgres Operator. As such, we need to validate that the Postgres Operator's configured
+		// image tag (first value) is compatible (i.e. is the same Major PostgreSQL version) as the
+		// existing cluster's PG value, and check that the OS values match.
+		if !upgradeTagValid(cl.Spec.CCPImageTag, apiserver.Pgo.Cluster.CCPImageTag) {
+			log.Debugf("Cannot upgrade from %s to %s. Image must be the same base OS and the upgrade must be within the same major PG version.", cl.Spec.CCPImageTag, apiserver.Pgo.Cluster.CCPImageTag)
 			response.Status.Code = msgs.Error
-			response.Status.Msg = "can not upgrade to the same image tag " + imageToUpgradeTo + " " + cl.Spec.CCPImageTag
+			response.Status.Msg = fmt.Sprintf("cannot upgrade from %s to %s, upgrade task failed.", cl.Spec.CCPImageTag, apiserver.Pgo.Cluster.CCPImageTag)
 			return response
 		}
-		log.Debugf("upgrading to image tag %s", imageToUpgradeTo)
-		spec.Parameters["CCPImageTag"] = imageToUpgradeTo
+
+		// given the above check passed, save the CCP image tag value
+		spec.Parameters["CCPImageTag"] = cl.Spec.CCPImageTag
 
 		// Create an instance of our CRD
 		err = kubeapi.Createpgtask(apiserver.RESTClient, newInstance, ns)
 		if err != nil {
 			response.Status.Code = msgs.Error
 			response.Status.Msg = err.Error()
+			response.WorkflowID = spec.Parameters[crv1.PgtaskWorkflowID]
 			return response
 		}
 
-		msg := "created minor upgrade task for " + clusterName
+		msg := "created upgrade task for " + clusterName
 		response.Results = append(response.Results, msg)
-
+		response.WorkflowID = spec.Parameters[crv1.PgtaskWorkflowID]
 	}
 
 	return response
 }
+
+// supportedOperatorVersion validates the Postgres Operator version
+// information for the candidate pgcluster. If this value is in the
+// required range, return true so that the upgrade may continue. Otherwise,
+// return false.
+func supportedOperatorVersion(version string) bool {
+	// get the Operator version
+	operatorVersionRegex := regexp.MustCompile(`^(\d)\.(\d)\.(\d)`)
+	operatorVersion := operatorVersionRegex.FindStringSubmatch(version)
+
+	// if this regex passes, the returned array should always contain
+	// 4 values. At 0, the full match, then 1-3 are the three defined groups
+	// If this is not true, the upgrade cannot continue (and we won't want to
+	// reference potentially missing array items).
+	if len(operatorVersion) != 4 {
+		return false
+	}
+
+	// if the first group does not equal the current major version
+	// then the upgrade cannot continue
+	major, err := strconv.Atoi(operatorVersion[1])
+	if err != nil {
+		log.Errorf("Cannot convert Postgres Operator's major version to an integer. Error: ", err)
+		return false
+	}
+	if major != REQUIRED_MAJOR_PGO_VERSION {
+		return false
+	}
```

Comment:
here's a trick:

```go
if major, err := strconv.Atoi(operatorVersion[1]); err != nil {
    log.Errorf("Cannot convert Postgres Operator's major version to an integer. Error: ", err)
    return false
} else if major != REQUIRED_MAJOR_PGO_VERSION {
    return false
}
```

Reply:
Also `log.Error` would be good enough here, the way you're using it.

Reply:
nice, updated with both changes. Thanks!